query-projector/src/lib.rs

```diff
@@ -243,23 +487,119 @@ fn project_elements<'a, I: IntoIterator<Item = &'a Element>>(
                 }
-            }
+            },
+            &Element::Aggregate(ref a) => {
+                if let Some(simple) = a.to_simple() {
+                    aggregated.insert(simple.var.clone());
+
+                    // When we encounter a simple aggregate -- one in which the aggregation can be
+                    // implemented in SQL, on a single variable -- we just push the SQL aggregation op.
+                    // We must ensure the following:
+                    // - There's a column for the var.
+                    // - The type of the var is known to be restricted to a sensible input set
+                    //   (not necessarily a single type, but e.g., all vals must be Double or Long).
+                    // - The type set must be appropriate for the operation. E.g., `Sum` is not a
+                    //   meaningful operation on instants.
+
+                    let (projected_column, return_type) = projected_column_for_simple_aggregate(&simple, &query.cc)?;
+                    cols.push(projected_column);
+
+                    // We might regret using the type tag here instead of the `ValueType`.
+                    templates.push(TypedIndex::Known(i, return_type.value_type_tag()));
+                    i += 1;
+                } else {
+                    // TODO: complex aggregates.
+                    bail!(ErrorKind::NotYetImplemented("complex aggregates".into()));
+                }
+            },
         }
     }
 
-    for var in with {
-        // We need to collect these into the SQL column list, but they don't affect projection.
-        // If a variable is of a non-fixed type, also project the type tag column, so we don't
-        // accidentally unify across types when considering uniqueness!
-        let (column, name) = candidate_column(&query.cc, &var);
+    // Anything we're projecting, or that's part of an aggregate, doesn't need to be in GROUP BY.
+    //
+    // Anything used in ORDER BY (which we're given in `named_projection`)
+    // needs to be in the SQL column list so we can refer to it by name.
+    //
+    // They don't affect projection.
+    //
+    // If a variable is of a non-fixed type, also project the type tag column, so we don't
+    // accidentally unify across types when considering uniqueness!
+    // Similarly, the type tag needs to be grouped.
+    // extra_projection: extras - variables
+    for var in query.named_projection.iter() {
+        if variables.contains(var) {
+            continue;
+        }
+
+        // If it's a fixed value, we need do nothing further.
+        if query.cc.is_value_bound(&var) {
+            continue;
+        }
+
+        let (column, name) = candidate_column(&query.cc, &var)?;
         cols.push(ProjectedColumn(column, name));
-        if query.cc.known_type(&var).is_none() {
+
+        // We don't care if a column has a single _type_, we care if it has a single type _code_,
```

Comment:
We have very carefully used "type tag" to be "numeric, Mentat-specific type encoding".  And we've called the `i32` type definition `ValueTypeTag`.  Let's not lose that!  s/code/tag/ everywhere.